---
title: "Diff/ial Expression analysis of CTRL & DM1-NT & DM1_sg2-DMPK"
author: "Maria Kondili"
date: "01/02/2022"
output: html_document
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}

suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(AnnotationDbi))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(pheatmap))
library(RColorBrewer)

```


```{r read_metadata}

data_dir <- "/dm1_vs_sgDMPK/HTSeq_counts/"
deseq_dir <- "/dm1_vs_sgDMPK/DESeq2_Analysis/"
metadata <- read_delim(paste0(deseq_dir,"metadata.tsv"),delim="\t",col_names=TRUE,show_col_types = FALSE)

metadata %>% glimpse
```

```{r create_counts_table}

for(i in 1:nrow(metadata)){

    f <- metadata$SampleID[i]
    sample <- gsub("_counts.txt","", f)
    ## remove other '_' in gene.names: (grep "_" $f)
    ## for f in ./*.tsv ; do  sed -i 's/Y_RNA/YRNA/g' $f;done
    ## for f in ./*.tsv ; do  sed -i 's/_.A/-A/g' $f;done
    tmp = readr::read_delim(paste0(data_dir,f),
                            comment = "_", delim="\t\t",
                            col_names = FALSE,
                            show_col_types = FALSE)
    colnames(tmp)  <- c("geneID",sample)

    if(i==1) { rawdata <- tmp; print(dim(rawdata)) }
    # 61533     2
    else rawdata <- cbind(rawdata,tmp[,2])
    # Note: Can do CBIND 2nd column with counts ,because 1st columns always same:geneID
}

rawdata %>% head()

### Remove Duplicate GeneIDs that have no Counts(NA) :
rawdata <- drop_na(rawdata)

rownames(rawdata) <- gsub("\\..*",  "", rawdata$geneID)

rawdata %>% dim

rawdata <- rawdata %>% dplyr::select(-c("geneID")) # keep rownames but ignore 1st col with geneID


```


```{r dotplot-counts}
##### BoxPlot of Raw-Counts per condition, to see outliers/ low-reads  ####

library(data.table)


mlt_counts <- reshape2::melt(as.data.table(rawdata),
                             value.name="Counts",
                             variable.name="Samples")
mlt_counts$conditions <- gsub("-[1-5]$","", mlt_counts$Samples )

ggplot(mlt_counts,aes(y=Counts,x=conditions,fill=conditions) ) +
      geom_dotplot(binaxis='y',
                   stackdir='center',
                   method = "dotdensity",
                   stackratio=1.5,
                   dotsize=1,na.rm = TRUE) # + scale_y_continuous (trans='log10')

```


```{r conversion_function}

.convert_ID2Name <- function(id_list){

  library(org.Hs.eg.db) # installed via biocLite.
                        #lib "org" hard-coded
  require(AnnotationDbi)

  gene_symbols <- mapIds(x=org.Hs.eg.db,
                         keys=id_list,
                         column="SYMBOL",
                         keytype="ENSEMBL",
                         multiVals="first")
  return(gene_symbols)

}

```


```{r filter_low_reads}

genes2keep <- which(rowSums(rawdata) >=30) # min.10 counts per condition
rawdata.filt <- rawdata[genes2keep,]

# rawdata.filt$geneName <- .convert_ID2Name(rownames(rawdata.filt))
# rawdata.filt <- relocate(rawdata.filt,"geneName")

readr::write_delim(rawdata.filt,"Results/Raw_Counts_all_conditions.tsv")

```


```{r colData_deseq }

conditions.dds <- gsub("-[1-5]$", "",colnames(rawdata.filt))
conditions.dds <- gsub("-","_" ,conditions.dds)
colData.dds <- data.frame(groups = conditions.dds )
rownames(colData.dds) <- colnames(rawdata.filt)

```


```{r matrix_deseqdataset }

dds.mat <- DESeqDataSetFromMatrix(rawdata.filt,
                                  colData=colData.dds,
                                  design=~groups)
# design formula must have column-name of colData ="groups"

dds.mat <- estimateSizeFactors(dds.mat)
norm_dds_counts <- counts(dds.mat, normalized=TRUE)
## DESeq is always applied in DDS object. Norm_counts will be used later
des <- DESeq(dds.mat)


```


```{r results_comparisons}

res.nt2 <- results(des,contrast=c("groups","DM_sgNT2" ,"CT_sgNT2"))
colnames(res.nt2)[-1] <- paste0(colnames(res.nt2)[-1],"_DM1vsCTRL_NT2")
res.nt2 %>% head

res.dm <- results(des,contrast=c("groups","DM_sgDMPK2","DM_sgNT2"))
colnames(res.dm)[-1] <- paste0(colnames(res.dm)[-1],"_sgDMPK_vs_sgNT")

res.sg2ct <- results(des,contrast=c("groups","DM_sgDMPK2","CT_sgNT2"))
colnames(res.sg2ct)[-1] <- paste0(colnames(res.sg2ct)[-1],"_sgDMPK_vs_CT")


```


```{r merge_res}

res_mrg <- cbind(as.data.frame(res.nt2),
                 as.data.frame(res.dm)[,-1],
                 as.data.frame(res.sg2ct)[,-1])

num_cols <- grep("baseMean|log2FoldChange_*|lfcSE_*|stat_*",colnames(res_mrg) )
names(num_cols) <- colnames(res_mrg)[num_cols]

##> apply "round" func with map_dfc > keep result in columns
res_mrg[,num_cols] <- map_dfc(.x=num_cols,  ~round(res_mrg[,.x],3))

res_mrg$GeneID <- rownames(res_mrg)
#res_mrg$GeneID <- gsub("\\..*",  "", res_mrg$GeneID)#> done before

res_mrg %>% glimpse
```


```{r add_counts_to_RES }

DE_tab <- as.data.frame(cbind(res_mrg, norm_dds_counts))


```


```{r convert_geneID}

rownames(norm_dds_counts) <- gsub("\\..*",  "",rownames(norm_dds_counts))
DE_tab$geneName <- .convert_ID2Name(DE_tab$GeneID)

#>Bring in front columns
DE_tab <- DE_tab %>% relocate(c("GeneID","geneName")) %>%
                     dplyr::select(-c("baseMean"))

DE_tab %>% glimpse
```

#####> Add chromosome Information to Results

```{r add_chr}

library(biomaRt)

ensembl <- useMart("ensembl", dataset="hsapiens_gene_ensembl")


ensembl_attrib <- getBM(attributes=c("ensembl_gene_id",
                                     "external_gene_name",
                                     "chromosome_name",
                                     "start_position",
                                     "end_position"), #>find which: listAttributes(ensembl)$name
                        filters="ensembl_gene_id",
                        values = DE_tab$GeneID,
                        mart=ensembl)

write_delim(ensembl_attrib,"ENS_GeneID_with_chr-start-stop_homoSapiens.tsv",
            col_names=T, delim="\t")

###  match lines of 2 tables by ID

matched_ids <- match(DE_tab$GeneID,ensembl_attrib$ensembl_gene_id)

ensembl_attrib <- ensembl_attrib[matched_ids,]

DE_tab$chr <- ensembl_attrib$chromosome_name

DE_tab <- DE_tab %>% relocate(c("GeneID","geneName", "chr"))

DE_tab$chr %>% as.factor %>% levels


```


```{r save_to_file}

out_dir <- paste0(deseq_dir,"Results/")

dir.create(out_dir,showWarnings = FALSE)

de_normCounts_filename = "Diff_Expression_and_NormCounts.tsv"

DE_tab <- DE_tab[order(DE_tab$padj_sgDMPK_vs_sgNT,decreasing=F),]

write_delim(DE_tab, paste0(out_dir,de_normCounts_filename),
            delim="\t",col_names = TRUE,na = "NA")

```


```{r save_raw_counts}

DE_and_raw <- as.data.frame(cbind(res_mrg, rawdata.filt))

##Add Ensembl-CHR:
matched_ids <- match(DE_and_raw$GeneID,ensembl_attrib$ensembl_gene_id)

ensembl_attrib <- ensembl_attrib[matched_ids,]

DE_and_raw$chr <- ensembl_attrib$chromosome_name

DE_and_raw$geneName <- ensembl_attrib$external_gene_name

DE_and_raw <- DE_and_raw %>% relocate(c("GeneID", "geneName", "chr"))
##DE_and_raw <- DE_and_raw %>% filter(chr != "Y")

de_rawCounts_filename  = "Diff_Expression_and_RawCounts.tsv"


write_delim(DE_and_raw, paste0(out_dir,de_rawCounts_filename),
            delim="\t",col_names = TRUE,na = "NA")

```

#### PCA

```{r pca }

plots_dir <- paste0(deseq_dir, "Plots/")
dir.create(plots_dir,showWarnings = F)

#png(paste0(plots_dir,"PCA_of_samples.png"))
rld <- rlog(des,blind=F)
DESeq2::plotPCA(rld,intgroup="groups") + ggtitle("PCA 3 conditions")
#dev.off()

```

#### HEATMAP

```{r func_heatmap }

#> FUNCTION :
create_heatmap <- function(expr_counts,design, subtitle) {

  calc_z_score <- function(x){
    (x - mean(x)) / sd(x)
  }

  Counts.z <- t(apply(expr_counts, 1, calc_z_score))

  sample_annot <- design


  library(RColorBrewer)
  p <- pheatmap(Counts.z,
           color = colorRampPalette(rev(brewer.pal(n=10,name ="RdYlBu")))(100),
           annotation_col=sample_annot,
           clustering_distance_rows="euclidean",
           cluster_cols=FALSE,
           annotation_legend = TRUE,
           annotation_names_col = TRUE,
           show_rownames=FALSE,
           labels_row = NULL,
           angle_col = 45,
           main = paste(nrow(Counts.z),"Diff/ly Expressed Genes\n(padj", subtitle,"<0.05)",sep=" "))
    print(p)

}

```

#### modified heatmap with hclust-obj & diff. distance_rows !euclidean

```{r}

create_heatmap <- function(expr_counts,design, subtitle) {
   library(RColorBrewer)
   library(pheatmap)
   calc_z_score <- function(x){
      (x - mean(x)) / sd(x)
    }

  Counts.z <- t(apply(expr_counts, 1, calc_z_score))

  ###> PRETTY-HEATMAP
  genes_hclust <- hclust(dist(Counts.z), method = "complete")

  # clust_assign <- cutree(tree = genes_hclust, k = 3)

  p <- pheatmap(Counts.z,
       color = colorRampPalette(rev(brewer.pal(n=10,name ="RdYlBu")))(100),
       annotation_col=design,
       cluster_rows=genes_hclust,
       cluster_cols=FALSE,
       annotation_legend = TRUE,
       annotation_names_col = TRUE,
       show_rownames = FALSE,
       labels_row = NULL,
       angle_col = 45,
       main = paste(nrow(Counts.z),"Diff/ly Expressed Genes\n(padj", subtitle,"<0.05)",sep=" "))
    print(p) #clustering_distance_rows="euclidean",

}

```

Define the Cut-off of log2FC for further graphs:

```{r l2fc_cutoff }

l2fc_co <- 0.5

```

 DM1 -vs- CTRL

```{r signif_DE_genes_DM1vsCTRL }

sig1.DE_tab <- subset(DE_tab, padj_DM1vsCTRL_NT2 < 0.05 & abs(log2FoldChange_DM1vsCTRL_NT2)>= l2fc_co)

print("\n Significant genes left: \n")
sig1.DE_tab %>% nrow()

sig1.norm_counts <- norm_dds_counts[which(rownames(norm_dds_counts) %in% sig1.DE_tab$GeneID),]

ht_outdir <- paste0("Plots/Heatmaps/log2fc=",l2fc_co)

png(paste0(ht_outdir,"/Heatmap_DM1-NT_vs_CTRL-NT.png"))
  create_heatmap(sig1.norm_counts,
                colData.dds,
                subtitle="DM1.NT-vs-CTRL.NT")
dev.off()


```

sgDMPK -vs- DM1.NT

```{r heatmap_dm1sg2_vs_dm1NT }

sig2.DE_tab <-  subset(DE_tab, padj_sgDMPK_vs_sgNT < 0.05 & abs(log2FoldChange_sgDMPK_vs_sgNT)>= l2fc_co )

print("\n Significant genes left: \n")
sig2.DE_tab %>% nrow() # 2032

sig2.norm_counts <- norm_dds_counts[which(rownames(norm_dds_counts) %in% sig2.DE_tab$GeneID),]

png(paste0(ht_outdir,"/Heatmap_sgDMPK2_vs_DM1-NT.png"))
create_heatmap(sig2.norm_counts,
               colData.dds,
               subtitle="DMsgDMPK_vs_DMsgNT")
dev.off()


```

sgDMPK -vs- CTRL

```{r heatmap_dm1sgDMPK_vs_CTsgNT }

sig3.DE_tab <-  subset(DE_tab, padj_sgDMPK_vs_CT < 0.05 & abs(log2FoldChange_sgDMPK_vs_CT)>= l2fc_co )

print("\n Significant genes left: \n")
sig3.DE_tab %>% nrow() # 6893

sig3.norm_counts <- norm_dds_counts[which(rownames(norm_dds_counts) %in% sig3.DE_tab$GeneID),]

png(paste0(ht_outdir,"/Heatmap_sgDMPK2_vs_CTRL.png"))
create_heatmap(sig3.norm_counts,
               colData.dds,
               subtitle="sgDMPK_vs_CTsgNT" )
dev.off()


```

####  Example of k-means heatmap

```{r}

expr_counts <- sig1.norm_counts
design <- colData.dds
Counts.z <- t(apply(sig1.norm_counts, 1, calc_z_score))
pheatmap(Counts.z,
          color = colorRampPalette(rev(brewer.pal(n=10,name ="RdYlBu")))(100),
          annotation_col=design,clustering_distance_rows ="euclidean",
          kmeans_k = 6,
          clustering_method ="median",
          cluster_cols=FALSE,
          annotation_legend = TRUE,
          annotation_names_col = TRUE,
          show_rownames = FALSE,
          labels_row = NULL,
          angle_col = 45)

```


####> PATHWAYS & GO-terms :


```{r function_barplot_GOterms}

source("create_barplot_GOterms_hs.R")

```

####> DM.sgNT-vs-CTRL.sgNT

```{r create_barplot_1 }

goterms_dir <- paste0("Plots/GO_terms/log2fc=", l2fc_co,"/")

###!! l2fc_co=1 -->re-create png

#> Biol.Processes
create_barplot_GOterms.hs(sig1.DE_tab.y$geneName,
                          gene_format = "SYMBOL",
                          my_ontol="BP",
                          plots_dir=goterms_dir,
                          main="DE-genes_DMsgNT-vs-CTsgNT")


##> Only Mol.Functions:
create_barplot_GOterms.hs(sig1.DE_tab.y$geneName,
                          gene_format = "SYMBOL",
                          my_ontol="MF",
                          plots_dir=goterms_dir,
                          main="DE-genes_DMsgNT-vs-CTsgNT")

```

####> DM.sgDMPK-vs-DM.NT

```{r create_barplot_2 }


##> Only Biol.Processes :
create_barplot_GOterms.hs(sig2.DE_tab.y$geneName,
                          gene_format = "SYMBOL",
                          my_ontol="BP",
                          plots_dir=goterms_dir ,
                          main="DE-genes_DM.sgDMPK-vs-DM.NT")

##> Only Mol.Functions:
create_barplot_GOterms.hs(sig2.DE_tab.y$geneName,gene_format = "SYMBOL",
                          my_ontol="MF",
                          plots_dir=goterms_dir,
                          main="DE-genes_DM.sgDMPK-vs-DM.NT")

```


####> DM.sgDMPK-vs-CTRL.sgNT

```{r create_barplot_3 }

create_barplot_GOterms.hs(sig3.DE_tab.y$geneName,gene_format = "SYMBOL",
                          my_ontol="BP",
                          plots_dir=goterms_dir,
                          main="DE-genes_DM.sgDMPK-vs-CTsgNT")

##> Only Mol.Functions:
create_barplot_GOterms.hs(sig3.DE_tab.y$geneName,gene_format = "SYMBOL",
                          my_ontol="MF",
                          plots_dir=goterms_dir,
                          main="DE-genes_DM.sgDMPK-vs-CTsgNT")

```

####> Add Correction % of profile of DM1 with sgDMPK :

```{r correction}

mean_expr_CTRL    <-  norm_dds_counts %>%
                      as.data.frame() %>%
                      dplyr::select(grep("CT-sgNT*",colnames(.))) %>% rowMeans()

mean_expr_DM1NT   <-  norm_dds_counts %>%
                      as.data.frame() %>%
                      dplyr::select(grep("DM-sgNT*",colnames(.))) %>% rowMeans()

mean_expr_DMsgDMPK <- norm_dds_counts %>%
                      as.data.frame() %>%
                      dplyr::select(grep("DM-sgDMPK*",colnames(.))) %>% rowMeans()


matched_rows <- match(rownames(norm_dds_counts), rownames(DE_tab))

DE_tab <- DE_tab[matched_rows,]
DE_tab$mean_expr_CTRL     <- mean_expr_CTRL
DE_tab$mean_expr_DM1NT    <- mean_expr_DM1NT
DE_tab$mean_expr_DMsgDMPK <- mean_expr_DMsgDMPK

##> Apply formula with mean-expr calculated
DE_tab <-  mutate(DE_tab,
                "Correction_Pct"= round( (1- ((mean_expr_DMsgDMPK - mean_expr_DM1NT)/
                                               (mean_expr_CTRL   - mean_expr_DM1NT)  )  ) *100, 2 )
                )

DE_tab[1:10,c("mean_expr_CTRL",
              "mean_expr_DM1NT",
              "mean_expr_DMsgDMPK",
              "Correction_Pct")]

write_delim(DE_tab, "Results/Diff_Expression_and_NormCounts_with_TreatmentCorrectionPercent.tsv",
             delim="\t", col_names=T)

```
